pyramid/tierFS.go

```diff
@@ -0,0 +1,143 @@
+package pyramid
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"time"
+
+	"github.com/treeverse/lakefs/block"
+)
+
+type storageProps struct {
+	t                  StorageType
+	localDir           string
+	blockStoragePrefix string
+}
+
+// ImmutableTierFS is a filesystem where written files are never edited.
+// All files are stored in the block storage. Local paths are treated as a
+// cache layer that will be evicted according to the given eviction algorithm.
+type ImmutableTierFS struct {
+	adaptor block.Adapter
+	config  map[StorageType]storageProps
+
+	// TODO: use refs anc last-access for the eviction algorithm
+	refCount   map[StorageType]map[string]int
+	lastAccess map[StorageType]map[string]time.Time
+}
+
+// mapping between supported storage types and their prefix
+var types = map[StorageType]string{
+	StorageTypeSSTable:      "sstables",
+	StorageTypeTreeManifest: "trees",
+}
+
+func NewTierFS(adaptor block.Adapter, localdir, blockStoragePrefix string) *ImmutableTierFS {
+	refCount := map[StorageType]map[string]int{}
+	lastAccess := map[StorageType]map[string]time.Time{}
+	config := map[StorageType]storageProps{}
+	for t, prefix := range types {
+		refCount[t] = map[string]int{}
+		lastAccess[t] = map[string]time.Time{}
+		config[t] = storageProps{
+			t:                  t,
+			localDir:           path.Join(localdir, prefix),
+			blockStoragePrefix: path.Join(blockStoragePrefix, prefix),
+		}
+	}
+
+	return &ImmutableTierFS{
+		adaptor:    adaptor,
+		refCount:   refCount,
+		lastAccess: lastAccess,
+		config:     config,
+	}
+}
+
+// Store uploads the local file to the block storage for persistence.
+func (tfs *ImmutableTierFS) Store(t StorageType, ns, filename string) error {
+	localpath := tfs.localpath(t, filename)
+	f, err := os.Open(localpath)
+	if err != nil {
+		return fmt.Errorf("open file: %w", err)
+	}
+	defer f.Close()
+
+	stat, err := f.Stat()
+	if err != nil {
+		return fmt.Errorf("file stat: %w", err)
+	}
+
+	return tfs.adaptor.Put(tfs.objPointer(t, ns, filename), stat.Size(), f, block.PutOpts{})
```

Comment:
Having to stat just to get the size of the file is really unfortunate -- e.g. it means we require the `Stat`.  (It can also be less safe than reading to the end.)

Reply:
Let me know if this comment is still valid assuming we must use files.